pennylane/fourier/spectrum.py

```diff
@@ -0,0 +1,271 @@
+
+        \sum \limits_{\omega_1\in \Omega_1} \dots \sum \limits_{\omega_N \in \Omega_N}
+        c_{\omega_1,\dots, \omega_N} e^{-i x_1 \omega_1} \dots e^{-i x_N \omega_N}
+
+    over the *frequency spectra* :math:`\Omega_j \subseteq \mathbb{R},`
+    :math:`j=1,\dots,N`. Each spectrum has the property that
+    :math:`0 \in \Omega_j`, and the spectrum is
+    symmetric (for every :math:`\omega \in \Omega_j` we have that :math:`-\omega \in
+    \Omega_j`). If all frequencies are integer-valued, the Fourier sum becomes a
+    *Fourier series*.
+
+    As shown in `Vidal and Theis (2019)
+    <https://arxiv.org/abs/1901.11434>`_ and `Schuld, Sweke and Meyer (2020)
+    <https://arxiv.org/abs/2008.08605>`_, if an input :math:`x_j, j = 1 \dots N`,
+    only enters into single-parameter gates of the form :math:`e^{-i x_j G}` (where :math:`G` is a Hermitian generator),
+    the frequency spectrum :math:`\Omega_j` is fully determined by the eigenvalues
+    of :math:`G`. In many situations, the spectra are limited
+    to a few frequencies only, which in turn limits the function class that the circuit
+    can express.
+
+    The ``spectrum`` function computes all frequencies that will potentially appear in the
+    sets :math:`\Omega_1` to :math:`\Omega_N` (which correspond to the :math:`N` different strings
+    used as an ``id`` to mark the input-encoding gates).
+
+    **Example**
+
+    Consider the following example, which uses non-trainable inputs ``x`` and
+    trainable parameters ``w`` as arguments to the qnode.
+
+    .. code-block:: python
+
+        import pennylane as qml
+        import numpy as np
+        from pennylane.fourier import spectrum
+
+        n_layers = 2
+        n_qubits = 3
+        dev = qml.device("default.qubit", wires=n_qubits)
+
+        @qml.qnode(dev)
+        def circuit(x, w):
+            for l in range(n_layers):
+                for i in range(n_qubits):
+                    qml.RX(x[i], wires=0, id="x"+str(i))
+                    qml.Rot(w[l,i,0], w[l,i,1], w[l,i,2], wires=0)
+            qml.RZ(x[0], wires=0, id="x0")
+            return qml.expval(qml.PauliZ(wires=0))
+
+        x = np.array([1, 2, 3])
+        w = np.random.random((n_layers, n_qubits, 3))
+        res = spectrum(circuit)(x, w)
+
+    >>> print(qml.draw(circuit)(x, w))
+    0: ──RX(1)──Rot(0.134, 0.601, 0.709)──RX(2)──Rot(0.0263, 0.0347, 0.446)──RX(3)──Rot(0.0606, 0.911, 0.366)
+    ──RX(1)──Rot(0.381, 0.398, 0.61)──RX(2)──Rot(0.956, 0.684, 0.0092)──RX(3)──Rot(0.538, 0.65, 0.732)──RZ(1)──┤ ⟨Z⟩
+
+    >>> for inp, freqs in res.items():
+    >>>     print(f"{inp}: {freqs}")
+    'x0': [-3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 3.0]
+    'x1': [-2.0, -1.0, 0.0, 1.0, 2.0]
+    'x2': [-2.0, -1.0, 0.0, 1.0, 2.0]
+
+    .. note::
+        While the Fourier spectrum usually does not depend
+        on trainable circuit parameters and the actual values of the inputs,
+        it may still change with arguments of the QNode that determine the architecture
```

Comment:
```suggestion
        on trainable circuit parameters or the actual values of the inputs,
        it may still change based on inputs to the QNode that alter the architecture
```